File: Network.java
public class Network extends AsyncTask<String, Integer, String> {

    OnAsyncRequestComplete caller;
    Context context;
    String method = "GET";
    List<NameValuePair> parameters = null;
    boolean handelResponse = true;

    // Three Constructors
    public Network(Activity a, String m, List<NameValuePair> p) {
        caller = (OnAsyncRequestComplete) a;
        context = a;
        method = m;
        parameters = p;
    }

    public Network(Activity a, String m, List<NameValuePair> p, boolean handled_reponse) {
        caller = (OnAsyncRequestComplete) a;
        context = a;
        method = m;
        parameters = p;
        handelResponse=  handled_reponse;
    }


    public Network(Activity a, String m) {
        caller = (OnAsyncRequestComplete) a;
        context = a;
        method = m;
    }

    public Network(Activity a) {
        caller = (OnAsyncRequestComplete) a;
        context = a;
    }

    // Interface to be implemented by calling activity
    public interface OnAsyncRequestComplete {
        public void asyncResponse(String response);
    }

    public String doInBackground(String... urls) {
        // get url pointing to entry point of API
        String address = urls[0].toString();
        if (method == "POST") {
            return post(address);
        }

        if (method == "GET") {
            return get(address);
        }

        return null;
    }

    public void onPreExecute() {

    }

    public void onProgressUpdate(Integer... progress) {
    }

    public void onPostExecute(String response) {
        if(handelResponse) {
            caller.asyncResponse(response);
        }else{
            Log.d("NETWORK_NHR_POST", response);
        }
    }

    protected void onCancelled(String response) {
        if(handelResponse) {
            caller.asyncResponse(response);
        }else{
            Log.d("NETWORK_NHR_CANCL", response);
        }
    }

    @SuppressWarnings("deprecation")
    private String get(String address) {
        try {

            if (parameters != null) {

                String query = "";
                String EQ = "="; String AMP = "&";
                for (NameValuePair param : parameters) {
                    query += param.getName() + EQ + URLEncoder.encode(param.getValue()) + AMP;
                }

                if (query != "") {
                    address += "?" + query;
                }
            }

            HttpClient client = new DefaultHttpClient();
            HttpGet get= new HttpGet(address);
            get.setHeader(new BasicHeader("Authorization", Utils.getInstance().getAccessToken()));


            HttpResponse response = client.execute(get);
            return stringifyResponse(response);

        } catch (ClientProtocolException e) {
            // TODO Auto-generated catch block
        } catch (IOException e) {
            // TODO Auto-generated catch block
        }

        return null;
    }

    private String post(String address) {
        try {

            HttpClient client = new DefaultHttpClient();
            HttpPost post = new HttpPost(address);
            post.setHeader(new BasicHeader("Authorization", Utils.getInstance().getAccessToken()));

            if (parameters != null) {
                post.setEntity(new UrlEncodedFormEntity(parameters));
            }

            HttpResponse response = client.execute(post);
            return stringifyResponse(response);

        } catch (ClientProtocolException e) {
            // TODO Auto-generated catch block
        } catch (IOException e) {
            // TODO Auto-generated catch block
        }

        return null;
    }

    private String stringifyResponse(HttpResponse response) {
        BufferedReader in;
        try {
            in = new BufferedReader(new InputStreamReader(response.getEntity().getContent()));

            StringBuffer sb = new StringBuffer("");
            String line = "";
            while ((line = in.readLine()) != null) {
                sb.append(line);
            }
            in.close();

            return sb.toString();
        } catch (IllegalStateException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        } catch (IOException e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }

        return null;
    }
}

File: Network.kt
import android.app.Activity
import android.os.AsyncTask
import android.os.Parcel
import android.os.Parcelable
import org.apache.http.NameValuePair
import java.net.HttpURLConnection
import java.net.URL
import java.io.*
import java.net.URLEncoder
import java.util.ArrayList


class Network() : AsyncTask<String, String, String>(), Parcelable {

    lateinit var requestType:String
    lateinit var activity: Activity
    lateinit var params:HashMap<String, String>
    lateinit var postData:StringBuilder
    lateinit var postDataBytes:ByteArray
    lateinit var onAsyncRequestComplete:OnAsyncRequestComplete

    constructor(parcel: Parcel) : this() {
        requestType = parcel.readString()
    }

    interface OnAsyncRequestComplete {
        fun asyncResponse(response:String)
    }

    constructor(activity:Activity, requestType:String, params: HashMap<String, String>) : this() {

        onAsyncRequestComplete = activity as OnAsyncRequestComplete;
        this.requestType = requestType
        this.activity = activity
        this.params = params
    }


    override fun onPreExecute() {
        // Before doInBackground

        postData = StringBuilder()
        for (param in params.entries) {
            if (postData.length != 0) postData.append('&')
            postData.append(URLEncoder.encode(param.key, "UTF-8"))
            postData.append('=')
            postData.append(URLEncoder.encode(param.value.toString(), "UTF-8"))
        }

        postDataBytes = postData.toString().toByteArray()
    }

    override fun doInBackground(vararg urls: String?): String {
        var urlConnection: HttpURLConnection? = null

        try {
            val url = URL(urls[0])

            urlConnection = url.openConnection() as HttpURLConnection
            urlConnection.setDoInput(true)
            if(requestType.equals("POST"))
                urlConnection.setDoOutput(true)
            urlConnection.requestMethod = requestType
            if(requestType.equals("POST"))
                urlConnection.getOutputStream().write(postDataBytes);

            var inString = streamToString(urlConnection.inputStream)

            publishProgress(inString)
        } catch (ex: Exception) {

        } finally {
            if (urlConnection != null) {
                urlConnection.disconnect()
            }
        }

        return " "
    }

    override fun onProgressUpdate(vararg values: String?) {
        onAsyncRequestComplete.asyncResponse(values[0].toString());
    }

    override fun onPostExecute(result: String?) {
        // Done
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(requestType)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Network> {
        override fun createFromParcel(parcel: Parcel): Network {
            return Network(parcel)
        }

        override fun newArray(size: Int): Array<Network?> {
            return arrayOfNulls(size)
        }
    }
}

fun streamToString(inputStream: InputStream): String {

    val bufferReader = BufferedReader(InputStreamReader(inputStream))
    var line: String
    var result = ""

    try {
        do {
            line = bufferReader.readLine()
            if (line != null) {
                result += line
            }
        } while (line != null)
        inputStream.close()
    } catch (ex: Exception) {

    }

    return result
}
